src/app/application.cpp

```diff
@@ -504,53 +504,19 @@ int Application::exec(const QStringList &params)
     connect(BitTorrent::Session::instance(), &BitTorrent::Session::torrentFinished, this, &Application::torrentFinished);
     connect(BitTorrent::Session::instance(), &BitTorrent::Session::allTorrentsFinished, this, &Application::allTorrentsFinished, Qt::QueuedConnection);
 
-#ifndef DISABLE_COUNTRIES_RESOLUTION
-    Net::GeoIPManager::initInstance();
-#endif
-    ScanFoldersModel::initInstance(this);
-
-#ifndef DISABLE_WEBUI
-    m_webui = new WebUI;
-#ifdef DISABLE_GUI
-    if (m_webui->isErrored())
-        return 1;
-    connect(m_webui, &WebUI::fatalError, this, []() { QCoreApplication::exit(1); });
-#endif // DISABLE_GUI
-#endif // DISABLE_WEBUI
-
-    new RSS::Session; // create RSS::Session singleton
-    new RSS::AutoDownloader; // create RSS::AutoDownloader singleton
-
-#ifdef DISABLE_GUI
-#ifndef DISABLE_WEBUI
-    Preferences* const pref = Preferences::instance();
-    // Display some information to the user
-    const QString mesg = QString("\n******** %1 ********\n").arg(tr("Information"))
-        + tr("To control qBittorrent, access the Web UI at %1")
-            .arg(QString("http://localhost:") + QString::number(pref->getWebUiPort())) + '\n'
-        + tr("The Web UI administrator user name is: %1").arg(pref->getWebUiUsername()) + '\n';
-    printf("%s", qUtf8Printable(mesg));
-    qDebug() << "Password:" << pref->getWebUiPassword();
-    if (pref->getWebUiPassword() == "f6fdffe48c908deb0f4c3bd36c032e72") {
-        const QString warning = tr("The Web UI administrator password is still the default one: %1").arg("adminadmin") + '\n'
-            + tr("This is a security risk, please consider changing your password from program preferences.") + '\n';
-        printf("%s", qUtf8Printable(warning));
-    }
-#endif // DISABLE_WEBUI
-#else
+#ifndef DISABLE_GUI
     m_window = new MainWindow;
 #endif // DISABLE_GUI
 
-    m_running = true;
+    // Slots are executed in the order of connection.
+    // Thus we need this connected last.
+    connect(BitTorrent::Session::instance(), &BitTorrent::Session::startupFinished, this, &Application::handleStartUpFinished);
+
+    m_paramsQueue = params + m_paramsQueue;
 
     // Now UI is ready to process signals from Session
     BitTorrent::Session::instance()->startUpTorrents();
```

Comment:
Is it still impossible to move it back to Session constructor?

Reply:
We don't want that. We have to have created the mainwindow (with its children) before we start emitting signals from the Session.

Reply:
But in your current implementation you don't emit any signals from startUpTorrents directly. They will be emitted wgen message loop is started, i.e. when main window is created.

Reply:
The commit that introduced this was 21a72c651f4439352942bec3573f0325b99765e9
The explanation in its commit message seems still relevant. If I am missing something, please enlighten me.

Reply:
>The explanation in its commit message seems still relevant. If I am missing something, please enlighten me.

I was quick to talk. Previously the UI responded immediately for every change during startup. Now it responds after the torrents have been loaded. So there's no chance of "losing" stuff.

What I don't understand is this: Why the fix worked? AFAIK, the mainloop doesn't start before `Base::exec()`. So the signals from the session wouldn't be tried to be delivered. So the mainwindow would have done its connects before Base::exec(). And after that the signals would have been delivered.

Reply:
>Why the fix worked?

Okay, let's leave it alone for now...